File: backend/src/algorithms/tsp.js
const { buildDistanceMatrix } = require('./distance');

/**
 * TSP Solver using Nearest Neighbor + Genetic Algorithm
 * 
 * Given N locations, find the shortest route visiting all of them.
 * The first location (index 0) is always fixed as the starting point.
 * Only the remaining locations are permuted during optimization.
 */
class TSPSolver {
  constructor(locations, options = {}) {
    this.locations = locations;
    this.n = locations.length;
    this.distMatrix = buildDistanceMatrix(locations);
    
    // GA parameters
    this.populationSize = options.populationSize || 60;
    this.generations = options.generations || 300;
    this.mutationRate = options.mutationRate || 0.15;
    this.eliteCount = options.eliteCount || 6;
  }

  // ─── Nearest Neighbor Heuristic ──────────────────────────────────
  // Always starts from index 0 (the fixed starting location)
  nearestNeighbor(startIndex = 0) {
    const visited = new Set([startIndex]);
    const path = [startIndex];
    let current = startIndex;

    while (visited.size < this.n) {
      let nearest = -1;
      let nearestDist = Infinity;

      for (let i = 0; i < this.n; i++) {
        if (!visited.has(i) && this.distMatrix[current][i] < nearestDist) {
          nearest = i;
          nearestDist = this.distMatrix[current][i];
        }
      }

      if (nearest === -1) break;
      visited.add(nearest);
      path.push(nearest);
      current = nearest;
    }

    return path;
  }

  // ─── Route total distance (full path including index 0) ─────────
  routeDistance(path) {
    let total = 0;
    for (let i = 0; i < path.length - 1; i++) {
      total += this.distMatrix[path[i]][path[i + 1]];
    }
    return total;
  }

  // ─── Genetic Algorithm ───────────────────────────────────────────
  // NOTE: The GA operates on "sub-routes" — permutations of indices
  //       1..n-1 only. Index 0 is always prepended as the start.

  // Create a random permutation of indices 1..n-1 (excluding 0)
  _randomSubRoute() {
    const route = Array.from({ length: this.n - 1 }, (_, i) => i + 1);
    // Fisher-Yates shuffle
    for (let i = route.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [route[i], route[j]] = [route[j], route[i]];
    }
    return route;
  }

  // Convert a sub-route to a full path (prepend 0)
  _toFullPath(subRoute) {
    return [0, ...subRoute];
  }

  // Convert a full path to a sub-route (remove leading 0)
  _toSubRoute(fullPath) {
    return fullPath.filter(idx => idx !== 0);
  }

  // Ordered Crossover (OX) — operates on sub-routes only
  _crossover(parent1, parent2) {
    const size = parent1.length;
    const start = Math.floor(Math.random() * size);
    const end = start + Math.floor(Math.random() * (size - start));

    const child = new Array(size).fill(-1);
    // Copy segment from parent1
    for (let i = start; i <= end; i++) {
      child[i] = parent1[i];
    }

    // Fill rest from parent2 in order
    let pos = (end + 1) % size;
    for (let i = 0; i < size; i++) {
      const candidate = parent2[(end + 1 + i) % size];
      if (!child.includes(candidate)) {
        child[pos] = candidate;
        pos = (pos + 1) % size;
      }
    }
    return child;
  }

  // Swap mutation — operates on sub-routes only
  _mutate(route) {
    if (Math.random() < this.mutationRate) {
      const i = Math.floor(Math.random() * route.length);
      const j = Math.floor(Math.random() * route.length);
      [route[i], route[j]] = [route[j], route[i]];
    }
    return route;
  }

  // 2-opt local improvement — operates on full path but skips index 0
  _twoOpt(route) {
    let improved = true;
    let bestRoute = [...route];
    let bestDist = this.routeDistance(bestRoute);

    while (improved) {
      improved = false;
      // Start from i=1 so we never move position 0 (the fixed start)
      for (let i = 1; i < bestRoute.length - 1; i++) {
        for (let j = i + 1; j < bestRoute.length; j++) {
          const newRoute = [...bestRoute];
          // Reverse the segment between i and j
          newRoute.splice(i, j - i + 1, ...bestRoute.slice(i, j + 1).reverse());
          const newDist = this.routeDistance(newRoute);
          if (newDist < bestDist) {
            bestRoute = newRoute;
            bestDist = newDist;
            improved = true;
          }
        }
      }
    }
    return bestRoute;
  }

  // Tournament selection — on sub-route population
  _tournamentSelect(population, fitnesses, tournamentSize = 5) {
    let best = -1;
    let bestFit = Infinity;
    for (let i = 0; i < tournamentSize; i++) {
      const idx = Math.floor(Math.random() * population.length);
      if (fitnesses[idx] < bestFit) {
        best = idx;
        bestFit = fitnesses[idx];
      }
    }
    return population[best];
  }

  // Full GA solve — all routes start at index 0
  geneticAlgorithm() {
    // Initialize population as sub-routes (indices 1..n-1 only)
    let population = [];

    // Seed with nearest neighbor starting from 0, then extract sub-route
    const nnFull = this.nearestNeighbor(0);
    population.push(this._toSubRoute(nnFull));

    // Add more seeds: nearest neighbor from different starts, but always
    // convert to sub-route format (remove the 0, it's always prepended)
    for (let i = 1; i < Math.min(this.n, this.eliteCount); i++) {
      const nn = this.nearestNeighbor(i);
      population.push(this._toSubRoute(nn));
    }

    // Fill rest with random sub-routes
    while (population.length < this.populationSize) {
      population.push(this._randomSubRoute());
    }

    let bestEverSub = null;
    let bestEverDist = Infinity;

    for (let gen = 0; gen < this.generations; gen++) {
      // Calculate fitness using full path (0 + sub-route)
      const fitnesses = population.map((sub) => this.routeDistance(this._toFullPath(sub)));

      // Track best
      for (let i = 0; i < population.length; i++) {
        if (fitnesses[i] < bestEverDist) {
          bestEverDist = fitnesses[i];
          bestEverSub = [...population[i]];
        }
      }

      // Sort by fitness and keep elites
      const sorted = population
        .map((route, idx) => ({ route, dist: fitnesses[idx] }))
        .sort((a, b) => a.dist - b.dist);

      const nextPopulation = sorted.slice(0, this.eliteCount).map((s) => s.route);

      // Generate offspring
      while (nextPopulation.length < this.populationSize) {
        const parent1 = this._tournamentSelect(population, fitnesses);
        const parent2 = this._tournamentSelect(population, fitnesses);
        let child = this._crossover(parent1, parent2);
        child = this._mutate(child);
        nextPopulation.push(child);
      }

      population = nextPopulation;
    }

    // Convert best sub-route back to full path
    let bestEver = bestEverSub ? this._toFullPath(bestEverSub) : [0];

    // Apply 2-opt local improvement to best solution (skips position 0)
    if (bestEver && bestEver.length <= 20) {
      bestEver = this._twoOpt(bestEver);
      bestEverDist = this.routeDistance(bestEver);
    }

    return {
      path: bestEver,
      totalDistanceKm: bestEverDist
    };
  }

  // ─── Main solve entry ────────────────────────────────────────────
  // Index 0 is always the first element in the returned path
  solve() {
    if (this.n <= 1) {
      return { path: [0], totalDistanceKm: 0 };
    }

    if (this.n === 2) {
      return { path: [0, 1], totalDistanceKm: this.distMatrix[0][1] };
    }

    if (this.n <= 3) {
      // For 3 nodes, use nearest neighbor starting from 0
      const path = this.nearestNeighbor(0);
      return { path, totalDistanceKm: this.routeDistance(path) };
    }

    // For 4+ nodes, use Genetic Algorithm (first location always fixed)
    return this.geneticAlgorithm();
  }
}

module.exports = TSPSolver;

File: backend/algorithms/tsp.py
import random
from .distance import build_distance_matrix

class TSPSolver:
    def __init__(self, locations, options=None):
        if options is None:
            options = {}
        self.locations = locations
        self.n = len(locations)
        self.dist_matrix = build_distance_matrix(locations)
        
        self.population_size = options.get('populationSize', 60)
        self.generations = options.get('generations', 300)
        self.mutation_rate = options.get('mutationRate', 0.15)
        self.elite_count = options.get('eliteCount', 6)

    def nearest_neighbor(self, start_index=0):
        visited = {start_index}
        path = [start_index]
        current = start_index

        while len(visited) < self.n:
            nearest = -1
            nearest_dist = float('inf')

            for i in range(self.n):
                if i not in visited and self.dist_matrix[current][i] < nearest_dist:
                    nearest = i
                    nearest_dist = self.dist_matrix[current][i]

            if nearest == -1:
                break
            visited.add(nearest)
            path.append(nearest)
            current = nearest

        return path

    def route_distance(self, path):
        total = 0.0
        for i in range(len(path) - 1):
            total += self.dist_matrix[path[i]][path[i + 1]]
        return total

    def _random_sub_route(self):
        route = list(range(1, self.n))
        random.shuffle(route)
        return route

    def _to_full_path(self, sub_route):
        return [0] + sub_route

    def _to_sub_route(self, full_path):
        return [idx for idx in full_path if idx != 0]

    def _crossover(self, parent1, parent2):
        size = len(parent1)
        start = random.randint(0, size - 1)
        if start == size - 1:
            end = start
        else:
            end = start + random.randint(0, size - start - 1)

        child = [-1] * size
        for i in range(start, end + 1):
            child[i] = parent1[i]

        pos = (end + 1) % size
        for i in range(size):
            candidate = parent2[(end + 1 + i) % size]
            if candidate not in child:
                child[pos] = candidate
                pos = (pos + 1) % size
        return child

    def _mutate(self, route):
        if random.random() < self.mutation_rate:
            i = random.randint(0, len(route) - 1)
            j = random.randint(0, len(route) - 1)
            route[i], route[j] = route[j], route[i]
        return route

    def _two_opt(self, route):
        improved = True
        best_route = list(route)
        best_dist = self.route_distance(best_route)

        while improved:
            improved = False
            for i in range(1, len(best_route) - 1):
                for j in range(i + 1, len(best_route)):
                    new_route = list(best_route)
                    # reverse the segment
                    new_route[i:j+1] = reversed(best_route[i:j+1])
                    new_dist = self.route_distance(new_route)
                    if new_dist < best_dist:
                        best_route = new_route
                        best_dist = new_dist
                        improved = True
        return best_route

    def _tournament_select(self, population, fitnesses, tournament_size=5):
        best = -1
        best_fit = float('inf')
        for _ in range(tournament_size):
            idx = random.randint(0, len(population) - 1)
            if fitnesses[idx] < best_fit:
                best = idx
                best_fit = fitnesses[idx]
        return population[best]

    def genetic_algorithm(self):
        population = []

        nn_full = self.nearest_neighbor(0)
        population.append(self._to_sub_route(nn_full))

        for i in range(1, min(self.n, self.elite_count)):
            nn = self.nearest_neighbor(i)
            population.append(self._to_sub_route(nn))

        while len(population) < self.population_size:
            population.append(self._random_sub_route())

        best_ever_sub = None
        best_ever_dist = float('inf')

        for gen in range(self.generations):
            fitnesses = [self.route_distance(self._to_full_path(sub)) for sub in population]

            for i in range(len(population)):
                if fitnesses[i] < best_ever_dist:
                    best_ever_dist = fitnesses[i]
                    best_ever_sub = list(population[i])

            sorted_pop = [x for _, x in sorted(zip(fitnesses, population))]
            
            next_population = [list(sub) for sub in sorted_pop[:self.elite_count]]

            while len(next_population) < self.population_size:
                parent1 = self._tournament_select(population, fitnesses)
                parent2 = self._tournament_select(population, fitnesses)
                child = self._crossover(parent1, parent2)
                child = self._mutate(child)
                next_population.append(child)

            population = next_population

        best_ever = self._to_full_path(best_ever_sub) if best_ever_sub else [0]

        if best_ever and len(best_ever) <= 20:
            best_ever = self._two_opt(best_ever)
            best_ever_dist = self.route_distance(best_ever)

        return {
            'path': best_ever,
            'totalDistanceKm': best_ever_dist
        }

    def solve(self):
        if self.n <= 1:
            return {'path': [0], 'totalDistanceKm': 0.0}

        if self.n == 2:
            return {'path': [0, 1], 'totalDistanceKm': self.dist_matrix[0][1]}

        if self.n <= 3:
            path = self.nearest_neighbor(0)
            return {'path': path, 'totalDistanceKm': self.route_distance(path)}

        return self.genetic_algorithm()
